test/mockgyver_tests.erl

```diff
@@ -387,9 +387,57 @@ removes_trace_pattern_test_aux() ->
 
     ?assertEqual([], flush()).
 
+%% A cached mocking module shall not be used if the original module
+%% has been changed.
+changed_module_is_used_for_mocking_mod_test_() ->
+    TCs = [create_dummy2_a,
+           mock_dummy2_a,
+           create_dummy2_b,
+           mock_dummy2_b],
+    ?WITH_MOCKED_SETUP(fun setup/0, fun cleanup/1,
+                       ?FOR_ALL_TIMEOUT, ?PER_TC_TIMEOUT,
+                       TCs).
+
+create_dummy2_a(_) ->
+    create_dummy2(a).
+
+mock_dummy2_a(_) ->
+    1 = mockgyver_dummy2:a(1),
+    ?assertError(undef, mockgyver_dummy2:b(1)),
+    ?WHEN(mockgyver_dummy2:a(1) -> 2),
```

Comment:
This may be a bit surprising :-), but the way mockgyver works is that it'll look for ?WHEN(mockgyver_dummy2, ...) in the entire test module and the resulting mocked module will have the union of all the functions.
I did this to avoid having to follow the call chain to get only the "right" functions in the test case. I was thinking it's probably ok to just look for all mocked functions in the entire module.

Anyway, we can try setting "OrigHash = foo" in mockgyver.erl.  By faking a non-changing hash we can see that the test case still passes...

The code loaded by create_dummy2/1 will immediately be unloaded by mockgyver before the next test case so we're unfortunately not getting that dummy2 in mock_dummy2.

I had to refresh my memory here... :-)

After some tinkering I've come up with another way of testing this which I hope will work. I tested it by setting "OrigHash = foo" and then the test case fails. Otherwise it passes.

```
%% A cached mocking module shall not be used if the original module
%% has been changed.
changed_module_is_used_instead_of_cached_test_() ->
    {timeout, ?PER_TC_TIMEOUT,
     fun changed_module_is_used_instead_of_cached_test_aux/0}.

changed_module_is_used_instead_of_cached_test_aux() ->
    create_dummy2(a),
    ?MOCK(fun() ->
                  %% fool mockgyver into mocking the module by mocking
                  %% a function that we will _not_ use within that module
                  ?WHEN(mockgyver_dummy2:c() -> ok),
                  1 = mockgyver_dummy2:a(1),
                  ?assertError(undef, mockgyver_dummy2:b(1))
          end),
    create_dummy2(b),
    ?MOCK(fun() ->
                  %% fool mockgyver into mocking the module by mocking
                  %% a function that we will _not_ use within that module
                  ?WHEN(mockgyver_dummy2:c() -> ok),
                  ?assertError(undef, mockgyver_dummy2:a(1)),
                  1 = mockgyver_dummy2:b(1)
          end).

```
I've also tested your change manually and it looks like it works fine.

Reply:
OK... Tricky!